RecoEgamma/EgammaTools/src/EgammaPCAHelper.cc

```diff
@@ -0,0 +1,342 @@
+    hitsize = rechitsFH.size();
+    for ( unsigned i=0; i< hitsize ; ++i) {
+        (*hitMap_)[rechitsFH[i].detid()] = & rechitsFH[i];
+    }
+    if (debug_)
+        std::cout << " FH " << hitsize << " RecHits " << std::endl;
+    hitsize = rechitsBH.size();
+    for ( unsigned i=0; i< hitsize ; ++i) {
+        (*hitMap_)[rechitsBH[i].detid()] = & rechitsBH[i];
+    }
+    if (debug_)
+        std::cout << " BH " << hitsize << " RecHits " << std::endl;
+    if( debug_)
+        std::cout << " Stored " << hitMap_->size() << " rechits " << std::endl;
+    pcaIteration_ = 0;
+    hitMapOrigin_ = 2;
+
+}
+
+void EGammaPCAHelper::storeRecHits(const reco::HGCalMultiCluster &cluster ){
+    theCluster_ = &cluster;
+    std::vector<std::pair<DetId, float>> result;
+    for (reco::HGCalMultiCluster::component_iterator it = cluster.begin(); it != cluster.end();
+    it++) {
+        const std::vector<std::pair<DetId, float>> &hf = (*it)->hitsAndFractions();
+        result.insert(result.end(),hf.begin(),hf.end());
+    }
+    storeRecHits(result);
+}
+
+void EGammaPCAHelper::storeRecHits(const reco::CaloCluster & cluster) {
+    theCluster_ = &cluster;
+    storeRecHits(cluster.hitsAndFractions());
+}
+
+void EGammaPCAHelper::storeRecHits(const std::vector<std::pair<DetId, float>> &hf) {
+    std::vector<double> pcavars;
+    pcavars.resize(3,0.);
+    theSpots_.clear();
+    pcaIteration_ = 0;
+
+    sigu_ = 0.;
+    sigv_ = 0.;
+    sigp_ = 0.;
+    sige_ = 0.;
+
+    unsigned hfsize = hf.size();
+    if (debug_)
+    std::cout << "The seed cluster constains " << hfsize << " hits " << std::endl;
+
+    if (hfsize == 0) return;
+
+
+    for (unsigned int j = 0; j < hfsize; j++) {
+        unsigned int layer = recHitTools_->getLayerWithOffset(hf[j].first);
+
+        const DetId rh_detid = hf[j].first;
+        std::map<DetId,const HGCRecHit *>::const_iterator itcheck= hitMap_->find(rh_detid);
+        if (itcheck == hitMap_->end()) {
+            std::cout << " Big problem, unable to find a hit " << rh_detid.rawId() << " " ;
+            std::cout << rh_detid.det() << " " << HGCalDetId(rh_detid) << std::endl;
+            continue;
+        }
+        if (debug_) {
+            std::cout << "DetId " << rh_detid.rawId() << " " << layer << " " <<  itcheck->second->energy() <<std::endl;
+            //        std::cout << " Hit " << itcheck->second << " " << itcheck->second->energy() << std::endl;
+        }
+        float fraction = hf[j].second;
+
+        double thickness =
+            (DetId::Forward == DetId(rh_detid).det()) ? recHitTools_->getSiThickness(rh_detid) : -1;
+        double mip = dEdXWeights_[layer] * 0.001;  // convert in GeV
+        if (thickness > 99. && thickness < 101)
+            mip *= invThicknessCorrection_[0];
+        else if (thickness > 199 && thickness < 201)
+            mip *= invThicknessCorrection_[1];
+        else if (thickness > 299 && thickness < 301)
```

Comment:
are these magic numbers known somewhere?
Please add some comments perhaps to also point to documentation

Reply:
These are provided by the HGCalRecProducers.HGCalRecHit_cfi and HGCalImagingAlgo.h for the maxlayer

Reply:
Added comments to their constructor.